Use ESC secrets

pgavlin opened the pull request:
These changes migrate this repo's GitHub Actions Workflows to use ESC secrets instead of GitHub Secrets.

The changes are largely mechanical:

- Common configuration for all ESC actions within a workflow is added to the workflow's environment variables
- Permissions are expanded as necessary for workflows that do not grant `id-token: write` permissions
	- `read-all` permissions are replaced with the union of all explicit read permissions and `id-token: write`
	- Default permissions are replaced with `write-all`, which is the equivalent of all explicit write permissions and
	  `id-token: write`
	- Explicit permissions are modified to grant `id-token: write`
- A step that fetches ESC secrets and populates environment variables is added to each step that reads secrets
- Direct references to secrets within the job are replaced with references to the step's outputs

All ESC actions are configured to fetch secrets from a shared ESC environment that contains secrets migrated from GitHub Actions. The ESC action performs its own OIDC exchange to obtain a Pulumi Access Token.

tgummerer commented on .github/workflows/stage-test.yml:
```diff
@@ -36,35 +37,37 @@ on:
-      ARM_SUBSCRIPTION_ID:   { required: false }
-      ARM_TENANT_ID:         { required: false }
-      CODECOV_TOKEN:         { required: false }
-
+      PULUMI_ACCESS_TOKEN: {required: false}
+      AWS_ACCESS_KEY_ID: {required: false}
+      AWS_SECRET_ACCESS_KEY: {required: false}
+      AWS_CI_ROLE_ARN: {required: false}
+      ARM_CLIENT_ID: {required: false}
+      ARM_CLIENT_SECRET: {required: false}
+      ARM_SUBSCRIPTION_ID: {required: false}
+      ARM_TENANT_ID: {required: false}
+      CODECOV_TOKEN: {required: false}
```
I think this whole section can now be deleted, instead of just changing the formatting?

pgavlin committed:
Use ESC secrets

These changes migrate this repo's GitHub Actions Workflows to use ESC secrets instead of GitHub Secrets.

The changes are largely mechanical:

- Common configuration for all ESC actions within a workflow is added to the workflow's environment variables
- Permissions are expanded as necessary for workflows that do not grant `id-token: write` permissions
	- `read-all` permissions are replaced with the union of all explicit read permissions and `id-token: write`
	- Default permissions are replaced with `write-all`, which is the equivalent of all explicit write permissions and
	  `id-token: write`
	- Explicit permissions are modified to grant `id-token: write`
- A step that fetches ESC secrets and populates environment variables is added to each step that reads secrets
- Direct references to secrets within the job are replaced with references to the step's outputs

All ESC actions are configured to fetch secrets from a shared ESC environment that contains secrets migrated from GitHub Actions. The ESC action performs its own OIDC exchange to obtain a Pulumi Access Token.

pgavlin commented:
Re-running CI.

pgavlin committed:
Source PULUMI_ACCESS_TOKEN from PULUMI_ACCESS_TOKEN_STAGING

stage-test.yml targets api.pulumi-staging.io; the org env's
PULUMI_ACCESS_TOKEN is not valid for that backend, but
PULUMI_ACCESS_TOKEN_STAGING is. Use the ESC export-rename syntax to expose
the staging token as PULUMI_ACCESS_TOKEN.
